### PR TITLE
[RFC][WIP] Mark output nodes with any post hook needed in disk_autograd engine

### LINKED ISSUE #33482: Distributed Data Parallel for models with remote parameter servers
## 🚀 Feature

The current [DDP] implementation (https://pytorch.org/tutorials/intermediate/ddp_tutorial.html) require a model is on a single node (maybe multiple devices). DDP won't work if a model has a remote parameter server and its `forward()` method makes RPCs to the remote parameter server. This feature is to make DDP work for models with parameter servers by leveraging the existing [Distributed Autograd/Optimizer](https://pytorch.org/docs/stable/notes/distributed_autograd.html).

Note that the parameter servers can't use DDP in this use case.

## Motivation

To support the following model

```
class SimpleNet(nn.Module):
    def __init__(self, d_in, d_out):
        super(SimpleNet, self).__init__()
        self.net = nn.Linear(d_in, d_out)
        self.relu = nn.ReLU()

    def forward(self, input):
        return self.relu(self.net(input))

class DdpModelWithRpc(nn.Module):
    def __init__(self, remote_parameter_server):
        super(DdpModelWithRpc, self).__init__()
        self.net1 = DDP(SimpleNet(5, 8))
        self.rps = remote_parameter_server
        self.rref = RPC call(remote_parameter_server, SimpleNet, 8, 5)
        self.net2 = DDP(SimpleNet(5, 3))

    def forward(self, x):
        x = self.net1(x)
        x = RPC(self.rps, SimpleNet.forward, x)
        return self.net2(x)

```

## Pitch

TDB

## Alternatives

TDB

## Additional context

CC: @pritamdamania87  @mrshenli @zhaojuanmao

### PR DESCRIPTION
**Pull request by hczhu**
# Summary

This is to make DistributedDataParallel work with modules which make RPCs in their forward.

A PyTorch use can wrap a module into DistributedDataParallel class regardless of whether the module make RPCs in its forward().
The backward pass will trigger all-reduce on the gradients of all local parameters (module.parameters()) the same way as how DDP currently works for a module.
Both the local parameters and remote parameters can be updated by an optimizer (either a vanila one or a distributed one).

No interface change at all. There are, however, some new requirements for the user code to make it work as expected.

- Both forward and backward passes have to be wrapping into with dist_autograd.context() if the module makes RPC calls in its forward method. dist_autograd.backward(context_id, [loss]) has to be used accordingly. (This is actually the same as it is now)
- A vanilla optimizer should be used to update the local parameters (possible with all-reduced gradients). A distributed optimizer should be used to update remote parameters. Don’t mix them together to update the same local parameter. (See implementation plan for the reason)

# Implementation plan

*Why doesn’t distributed autograd work with DDP now?*

1. dist_autograd marks all output nodes (module parameters usually) *unneeded.* That implies
    1. Their backward functions (usually AccumulateGrad) are not called in the autograd engine.
    2. The post hook, which does all-reduce, installed in each output nodes are not called accordingly.
2. The accenstor nodes of the output nodes are indeed evaluated. Therefore, the autograd engine does generate gradients for the output nodes. It’s just the gradients are not consumed by AccumulateGrad of the output nodes. Therefore, the tensors(local parameters) of output nodes don’t have local .grad, which DDP post hooks do all-reduce on. Instead, distributed autograd engine installs a callback on the gradients and stores them in a dist context global map. Note that the autograd engine is not aware of the map at all.

*What changes needed to make it work?*
(this is a proposal targeting minimum changes)

1. dist_autograd marks all output node(corresponding to local parameters), which has at one post hook, *needed* so that their .grad will be set. DPP post hooks will be called accordingly. That means the .grad of all local parameters will have the aggregated values after the backward pass.
2. dist_autograd keeps its callback the same as now so that the dist context gradient map will be populated as well. (no change here)

# Test Plan:

The existing tests.
Will add unit test for the changed part.

### CODE REVIEW

**Review by pritamdamania87 — COMMENTED** [review]
I guess we will update this based on the latest design?

**Comment by hczhu** [comment]
> I guess we will update this based on the latest design?

Yes, will update this PR shortly.

**Comment by hczhu** [comment]
Moved to https://github.com/pytorch/pytorch/pull/34501
